tests/test_server.py

```diff
@@ -0,0 +1,691 @@
+                "messages": [
+                    {"role": "user", "content": "What is the weather in Istanbul?"}
+                ],
+                "call_mode": "tools",
+                "code_interpreter": False,
+                "choice": "auto",
+            },
+            {
+                "test_aim": 'Single function_call with "auto"',
+                "messages": [
+                    {"role": "user", "content": "What is the weather in Istanbul?"}
+                ],
+                "call_mode": "functions",
+                "code_interpreter": False,
+                "choice": "auto",
+            },
+            {
+                "test_aim": 'Parallel tool_calls with "auto"',
+                "messages": [
+                    {
+                        "role": "user",
+                        "content": "What is the weather in Istanbul and Singapore respectively?",
+                    }
+                ],
+                "call_mode": "tools",
+                "code_interpreter": False,
+                "choice": "auto",
+            },
+            {
+                "test_aim": 'Parallel function_calls with "auto"',
+                "messages": [
+                    {
+                        "role": "user",
+                        "content": "What is the weather in Istanbul and Singapore respectively?",
+                    }
+                ],
+                "call_mode": "functions",
+                "code_interpreter": False,
+                "choice": "auto",
+            },
+            {
+                "test_aim": 'Normal text gen + tool_calls with "auto"',
+                "messages": {
+                    "openai": [
+                        {
+                            "role": "user",
+                            "content": "What is the weather in Istanbul? Answer this question: 'How are you?', before checking the weather.",
+                        }
+                    ],
+                    "meetkai/functionary-small-v2.4": [
+                        {
+                            "role": "user",
+                            "content": "Answer both these questions: 'How are you?' and 'What's the weather in Istanbul?'",
+                        }
+                    ],
+                    "meetkai/functionary-small-v2.5": [
+                        {
+                            "role": "user",
+                            "content": "What is the weather in Istanbul? Answer this question: 'How are you?', before checking the weather.",
+                        }
+                    ],
+                    "meetkai/functionary-small-v3.1": [
+                        {
+                            "role": "user",
+                            "content": "Answer both these questions: 'How are you?' and 'What's the weather in Istanbul?'",
+                        }
+                    ],
+                    "meetkai/functionary-small-v3.2": [
```

Comment:
So the test case depends on the model? I mean when we have a new model trained, we need to test to find out a case that the response contains both text & function calls ?

Reply:
It will be part of the unittest implementation for that model. Ideally, the new models should be more robust and are able to work with the test case for older models. If the new model does not support features like code generation, we can remove those tests in the main test method.